### PR TITLE
Typechecking edge weights of input graphs to ``Lattice``

### LINKED ISSUE #466: Type check edge weights in ``Lattice`` constructions?
### What is the expected enhancement?

Right now, a ``Lattice`` can be constructed from a graph with any value for the edge weights, e.g.
```python
>>> from qiskit_nature.problems.second_quantization.lattice import Lattice
>>> import retworkx as rx
>>> graph = rx.PyGraph(multigraph=False)
>>> graph.add_nodes_from(range(2))
>>> graph.add_edges_from([(0, 1, "dilution fridge")])
>>> Lattice(graph)
```
In the above case, the construction of a second quantized op or a qubit op based on the lattice will obviously fail. But there might be more subtle cases.

It would be nice to type check the weights and ensure that they are numeric (or in some supported format). Or is there a reason to support arbitrary types as weights?

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Closes #466.

### Details and comments

Split 1/2 of PR #467.

### CODE REVIEW

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py [hunk 1]
```diff
@@ -110,20 +113,27 @@ class Lattice:
     def __init__(self, graph: PyGraph) -> None:
         """
         Args:
-            graph: Input graph for Lattice. `graph.multigraph` must be False.
+            graph: Input graph for Lattice.
```
Comment (mrossinek) [inline qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:116]: Why this change?

File: releasenotes/notes/lattice-edge-typecheck-3f5b4896fe26c8dc.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+---
+fixes:
+  - |
+    Add a typecheck for the input graphs to
```
Comment (mrossinek) [inline releasenotes/notes/lattice-edge-typecheck-3f5b4896fe26c8dc.yaml:4]: `make spell` fails because of "typecheck"

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py [hunk 3]
```diff
@@ -110,20 +113,27 @@ class Lattice:
...
                 f"Invalid `graph.multigraph` {graph.multigraph} is given. "
                 "`graph.multigraph` must be `False`."
             )
-        if graph.edges() == [None] * graph.num_edges():
-            weighted_edges = [edge + (1.0,) for edge in graph.edge_list()]
-            for start, end, weight in weighted_edges:
-                graph.update_edge(start, end, weight)
+
+        # validate the edge weights
+        for edge_index, edge in graph.edge_index_map().items():
```
Comment (mrossinek) [inline qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:129]: I'm not familiar with retworkx but I see more changes than just adding the type check here. Can you shortly explain the differences and benefits of your implementation?

**Review by mrossinek — COMMENTED** [review]
Just some minor comments.

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py (hunk 3 above)
Comment (mrossinek) [reply qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:129]: I just checked out the #472 and I assume that you changed this in order to support both, retworkx and networkx.
If that's the case then it's all good :+1:

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py (hunk 1 above)
Comment (mrossinek) [reply qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:116]: I do see that this gets updated further in #472 but that does not mean we can remove this information in this PR, which will potentially be included in `0.3.1` whereas the other PR can only go into `0.4.0`.

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py (hunk 3 above)
Comment (Cryoris) [reply qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:129]: Yes, partially due to #472 but also to support the case where a single edge weight is `None` but others are not

File: qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py (hunk 1 above)
Comment (Cryoris) [reply qiskit_nature/problems/second_quantization/lattice/lattices/lattice.py:116]: Whoops, yes it seems I didn't split the PR properly 😉

**Review by mrossinek — DISMISSED** [review]
Looks good to me, thanks @Cryoris!
We'll probably hold off the merge until one other dev has the chance to review this but most of us are already on vacaiton.

File: .pylintdict [hunk 4]
```diff
@@ -265,6 +265,7 @@ mol
 morse
 mpl
 mul
+multigraph
```
Comment (dlasecki) [inline .pylintdict:268]: Inserted not in the alphabetical order.

**Review by dlasecki — CHANGES_REQUESTED** [review]

File: .pylintdict (hunk 4 above)
Comment (Cryoris) [reply .pylintdict]: Fixed in 5eb8e5b 🙂

**Review by mrossinek — APPROVED** [review]
LGTM, thanks!

**Review by dlasecki — APPROVED** [review]

**Review by pbark — APPROVED** [review]

**Comment by mrossinek** [comment]
Something appears to have gone wrong with mergify here. The PR is merged properly into `main` (see https://github.com/Qiskit/qiskit-nature/commit/aa910c9e8fda690535c9a1634b91c10cf0c682db) so I am closing this manually.
